README.md
# Avetmiss

This library is designed to help with the generation of AVETMISS NAT files.

[![Build Status](https://travis-ci.org/bluedogtraining/avetmiss.png?branch=master)](https://travis-ci.org/bluedogtraining/avetmiss)

## Install

Via Composer

``` bash
$ composer require bluedogtraining/avetmiss
```

## Usage

The idea behind the library is very simple.

1. Create fieldset with field definitions (or use a bundled fieldset)
2. Initiate a file with the fieldset.
3. Create a row from the file.
4. Populate the row.
5. Add the populated row back to the file.
6. Export the file.


```php
use Bdt\Avetmiss\File;
use Bdt\Avetmiss\Nat\V7\Nat120;

// array of student courses, pulled from the database
$studentcourses = DB::getStudentCourses();

// initiate a new nat file
$file = new File(new Nat120);

// loop through the studentcourses and add them to the file
foreach($this->studentcourses as $studentcourse) {

    try {
        $row = $file->createRow();
        
        $row->set('client_id', $studentcourse->Student->id);
        $row->set('subject_id', $studentcourse->Course->id);
        //...
        
        $file->writeRow($row);
    } catch(Exception $e) {
        // Display or log any errors.
    }
}

$file->export('nat120.txt');
```

### Extending

The library comes with preliminary Fieldset definitions for AVETMISS Version 7 NAT files.

You can very easily add your own NAT files if required.

```php
$myFieldset = new Fieldset([
    Field::make('date')->name('enrolment_date')->length(8),
    Field::make('numeric')->name('state_id')->length(2)->pad('0')->in(Config::keys('states')),
]);
```

Or own rules.

```php
class MyOwnConfig extends Config
{
    protected static $deliveryTypes = [
        10 => 'Classroom-based',
        20 => 'Electronic based',
        30 => 'Employment based',
        40 => 'Other delivery (eg correspondence)',
        90 => 'Not applicable - recognition of prior learning/ recognition of current competency/ credit transfer'
    ];
}
```

### Framework Integration

#### Laravel 5

This library comes with a service provider to add rules for validating against
AVETMISS NAT fields.

To use this, first add the `Bdt\Avetmiss\Frameworks\Laravel\ValidationServiceProvider` to the `providers` array in `config/app.php`.

```php
$validator = Validator::make([
    'my_start_date' => '01022000'
], [
    'my_start_date' => 'avetmiss:nat120,activity_start_date'
]);

$isValid = $validator->passes();
```

You can optionally pass a third boolean parameter to the `avetmiss` rule to enforce a maximum string length.

#### Zend Framework 1

This library comes with a utility for creating Zend Framework 1 validators based
on AVETMISS NAT fields.

```php
$factory = new Bdt\Avetmiss\Frameworks\Zf1\ValidatorFactory;
$validator = $factory->create('nat120', 'activity_start_date');
$validator->isValid('my_start_date');
```

You can optionally pass a third boolean parameter to the `ValidatorFactory::create` method to enforce a maximum string length.

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Attribution

This library is based upon information provided in the 
[AVETMISS VET Provider Collection Specifications](http://www.ncver.edu.au/wps/wcm/connect/dd52c92e-46c2-4217-9a21-6135c405f482/Avetmiss-r7-vet-provider-specifications-2578.pdf), 
which is provided under a [Creative Commons Attribution 3.0 Australia license](http://creativecommons.org/licenses/by/3.0/au/).



CHANGELOG.md
# CHANGELOG

## 4.5.0 - 2019-11-29

* Add support for middle name to nameForEncryption utility
* Update PHP version to 5.4 (package already contained 5.3 breaking short array syntax '[ ]' )

## 4.4.0 - 2019-10-24

* Support Laravel versions 6 and 7

## 3.0.2 - 2015-07-09

* Add NAT file descriptions
* Add CC attribution

## 3.0.1 - 2015-07-01

* Fix `File::writeRow` not appending to data.

## 3.0.0 - 2015-07-01

### Changed

* Namespace has been changed from `Avetmiss` to `Bdt\Avetmiss`
* `Row` is no longer an abstract class, and accepts a Fieldset on initialisation.
* `Field::isValid()` is now `Field::validate($value)`
* `Field::render()` is now `Field::render($value)`
* `Nat\V7\*` classes are now instances of `Fieldset`, to create a new Row, call
  `new Row(new Nat\V7\Nat120)` for example.
* `FormatInterface::isFormatValid()` (and implementing classes) now expect a `$value` argument.
* `Row::getField` now throws `Bdt\Avetmiss\Exceptions\FieldNotFoundException` instead of `UnexistingFieldException`
* `Row::populateFields` now throws `Bdt\Avetmiss\Exceptions\EmptyRowException` instead of `UnexistingFieldException`
* `Field::` methods `name`, `length`, `in`, and `pad` now return a new instance of Field instead of modifying the existing instance.
* `Config` is now an abstract class.
* `Field` protected values `name`, `length`, `in` and `pad` now have implicit `null` values, instead of defined.
* `FormatInterface` has been removed, its `isFormatValue` is now an abstract method on the abstract `Field` class.
* Fix typos in:
  * Nat010: `email_adress`
  * Nat080: `adress_location_suburb_locality_or_town`
  * Nat080: `adress_street_number`
  * Nat080: `adress_street_name`
* `File::addRow` is now `File::writeRow` and stores the rendered row string in memory, so that the row can be discarded.

### Added

* Added `Fieldset` class.
* Added `File::createRow` method.
* Added `Row::set` and `Row::get` methods. For performance, these should be favoured over the magic `__get` and `__set` methods.
* Added `Frameworks` directory with framework-specific utilities.
* Added `Row::render` method

### Removed

* Removed `Field::getValue()`, instead call `Row::__get($fieldName)`
* Removed `Field::setValue($value)`, instead call `Row::__set($fieldName, $value)`
* Removed `Row::getField()`, instead use `Row::getFieldset()->getFieldByName()`
* Removed `Row::addField()`, fields should now be added directly to the fieldset.
* Removed `Row::addFields()`, fields should now be added directly to the fieldset.

## 2.0.0 - 2014-03-24

* Support for PHP 5.3 removed.

## 1.0.0 - 2014-03-24

* Initial release.

src/Config/V8.php
<?php

namespace Bdt\Avetmiss\Config;

/**
 * Class for AVETMISS Version 7 options.
 */
class V8 extends Config
{

    /**
     * @var array
     */
    protected static $recognitionStatus = [
        '11' => 'Nationally accredited qualification designed to lead to a qualification specified in a national Training Package',
        '12' => 'Nationally recognised accredited course, other than a qualification designed to lead to a qualification specified in a national Training Package',
        '14' => 'Other courses'
    ];

    /**
     * @var array
     */
    protected static $qualificationCategories = [
        '211' => 'Graduate Diploma',
        '213' => 'Professional Specialist Qualification at Graduate Diploma Level',
        '221' => 'Graduate Certificate',
        '222' => 'Professional Specialist Qualification at Graduate Certificate Level',
        '311' => 'Bachelor Degree (Honours)',
        '312' => 'Bachelor Degree (Pass)',
        '411' => 'Advanced Diploma',
        '413' => 'Associate Degree',
        '421' => 'Diploma',
        '511' => 'Certificate IV',
        '514' => 'Certificate III',
        '521' => 'Certificate II',
        '524' => 'Certificate I',
        '611' => 'Year 12',
        '613' => 'Year 11',
        '621' => 'Year 10',
        '912' => 'Other Non-award Courses',
        '991' => 'Statement of Attainment Not Identifiable by Level',
        '992' => 'Bridging and Enabling Courses Not Identifiable by Level',
        '999' => 'Education not elsewhere classified'
    ];

    /**
     * @var array
     */
    protected static $fundingSourceNational = [
        '11' => 'Commonwealth and state general purpose recurrent',
        '13' => 'Commonwealth specific purpose programs',
        '15' => 'State specific purpose programs',
        '20' => 'Domestic full fee-paying client',
        '30' => 'International full fee-paying client',
        '80' => 'Revenue earned from another registered training organisation'
    ];

    /**
     * @var array
     */
    protected static $commencingCourseOptions = [
        '3' => 'Commencing enrolment in the qualification or course',
        '4' => 'Continuing enrolment in the qualification or course from a previous year',
        '8' => 'Unit of competency or module enrolment only'
    ];

    /**
     * @var array
     */
    protected static $vetFlagOptions = [
        'Y' => 'Yes - The intention of the program of study is vocational',
        'N' => 'No - The intention of the program of study is not vocational'
    ];

    /**
     * @var array
     */
    protected static $predominantDeliveryModes = [
        'I' => 'Internal delivery (e.g. classroom based)',
        'E' => 'External delivery (e.g. online)',
        'W' => 'Workplace based delivery',
        'N' => 'Not applicable (e.g. RPL & CT)'
        ];

    /**
     * @var array
     */
    protected static $deliveryModeIdentifiers = [
        'YNN' => 'Internal only',
        'NYN' => 'External only',
        'NNY' => 'Workplace based only',
        'YYN' => 'Combination of internal and external',
        'YNY' => 'Combination of internal and workplace based',
        'NYY' => 'Combination of external and workplace based',
        'YYY' => 'Combination of all modes',
        'NNN' => 'Not applicable (RPL or credit transfer)'
    ];

    /**
     * @var array
     */
    protected static $deliveryTypes = [
        '10' => 'Classroom-based',
        '20' => 'Electronic based',
        '30' => 'Employment based',
        '40' => 'Other delivery (eg correspondence)',
        '90' => 'Not applicable - recognition of prior learning/ recognition of current competency/ credit transfer'
    ];

    /**
     * @var array
     */
    protected static $deliveryTypeNotApplicableOutcomes = [
        51,
        52,
        53,
        54,
        60,
        // Commenting these out because we're in an interum version, not quite in V8 yet but this change is currently active for V7
        // 81,
        // 82
    ];

    /**
     * @var array
     */
    protected static $outcomeStatusOptions = [
        20 => 'Competency achieved/pass',
        30 => 'Competency not achieved/fail',
        40 => 'Withdrawn',
        51 => 'Recognition of prior learning - granted',
        52 => 'Recognition of prior learning - not granted',
        53 => 'Recognition of current competency - granted',
        54 => 'Recognition of current competency - not granted',
        60 => 'Credit transfer',
        65 => 'Gap Training',
        70 => 'Continuing enrolment',
        81 => 'Non-assessed enrolment - Satisfactorily completed',
        82 => 'Non-assessed enrolment - Withdrawn or not satisfactorily completed',
    ];

    /**
     * @var array
     */
    protected static $outcomeStatusCodes = [
        20,
        30,
        40,
        51,
        52,
        53,
        54,
        60,
        70,
        81,
        82,
        90
    ];

    /**
     * @var array
     */
    protected static $noCompletionWithOutcomeStatus = [
        30,
        40,
        52,
        54
    ];

    /**
     * @var array
     */
    protected static $studyReasonOptions = [
        '01' => 'To get a job',
        '02' => 'To develop my existing business',
        '03' => 'To start my own business',
        '04' => 'To try for a different career',
        '05' => 'To get a better job or promotion',
        '06' => 'It was a requirement of my job',
        '07' => 'I wanted extra skills for my job',
        '08' => 'To get into another course of study',
        '11' => 'Other reasons',
        '12' => 'For personal interest or self-development'
    ];

    /**
     * @var array
     */
    protected static $states = [
        '01' => 'New South Wales',
        '02' => 'Victoria',
        '03' => 'Queensland',
        '04' => 'South Australia',
        '05' => 'Western Australia',
        '06' => 'Tasmania',
        '07' => 'Northern Territory',
        '08' => 'Australian Capital Territory',
        '09' => 'Other Australian Territories or Dependencies',
        '99' => 'Other (Overseas but not an Australian Territory or Dependency)'
    ];

    /**
     * @var array
     */
    protected static $statesShort = [
        '01' => 'NSW',
        '02' => 'VIC',
        '03' => 'QLD',
        '04' => 'SA',
        '05' => 'WA',
        '06' => 'TAS',
        '07' => 'NT',
        '08' => 'ACT',
        '09' => 'Other Australian Territories or Dependencies',
        '99' => 'Other (Overseas but not an Australian Territory or Dependency)'
    ];

    /**
     * @var array
     */
    protected static $clientTitles = [
        'Mr'   => 'Mr',
        'Mrs'  => 'Mrs',
        'Miss' => 'Miss',
        'Ms'   => 'Ms',
        'Dr'   => 'Dr',
        'Rev'  => 'Rev',
        'Hon'  => 'Hon',
        'Sir'  => 'Sir',
        'Mx'   => 'Mx'
    ];

    /**
     * @var array
     */
    protected static $schoolTypes = [
        '21'   => 'School - Government',
        '25'  => 'School - Catholic',
        '27' => 'School - Independent',
        '31'   => 'Technical and Further Education institute',
        '61'   => 'Community based adult education provider',
        '91'  => 'Privately operated registered training organisation',
        '92'  => 'Home school arrangement'
    ];

    /**
     * @var array
     */
    protected static $surveyContactStatuses = [
        'A'   => 'Available for survey use',
        'C' => 'Correctional facility (address or enrolment)',
        'D'   => 'Deceased student',
        'E'   => 'Excluded from survey use',
        'I'  => 'Invalid address/itinerant student (very low likelihood of response)',
        'M'  => 'Minor - Under age of 15 (Not to be surveyed)',
        'O' => 'Overseas (address or enrolment)'
    ];

    /**
     * @var array
     */
    protected static $clientSex = [
        'M' => 'Male',
        'F' => 'Female',
        'X' => 'Other'
    ];

    /**
     * @var array
     */
    protected static $booleans = [
        'Y' => 'Yes',
        'N' => 'No'
    ];

    /**
     * @var array
     */
    protected static $booleanOptions = [
        '1' => 'Yes',
        '0' => 'No'
    ];

    /**
     * @var array
     */
    protected static $atSchool = [
        'Y' => 'Yes - the client is still attending secondary school',
        'N' => 'No - the client is not attending secondary school'
    ];

    /**
     * @var array
     */
    protected static $highestSchoolLevelCompleted = [
        '02' => 'Did not go to school',
        '08' => 'Year 8 or below',
        '09' => 'Year 9 or equivalent',
        '10' => 'Completed Year 10',
        '11' => 'Completed Year 11',
        '12' => 'Completed Year 12'
    ];

    /**
     * @var array
     */
    protected static $priorEducationAchievements = [
        '008' => 'Bachelor Degree or Higher Degree level (defined for AVETMISS use only)',
        '410' => 'Advanced Diploma or Associate Degree Level',
        '420' => 'Diploma Level',
        '511' => 'Certificate IV',
        '514' => 'Certificate III',
        '521' => 'Certificate II',
        '524' => 'Certificate I',
        '990' => 'Miscellaneous Education'
    ];

    /**
     * @var array
     */
    protected static $labourForceStatus = [
        '01' => 'Full-time employee',
        '02' => 'Part-time employee',
        '03' => 'Self-employed - not employing others',
        '04' => 'Employer',
        '05' => 'Employed - unpaid worker in a family business',
        '06' => 'Unemployed - seeking full-time work',
        '07' => 'Unemployed - seeking part-time work',
        '08' => 'Not employed - not seeking employment'
    ];

    /**
     * @var array
     */
    protected static $clientIndigenousStatus = [
        '1' => 'Yes, Aboriginal',
        '2' => 'Yes, Torres Strait Islander',
        '3' => 'Yes, Aboriginal AND Torres Strait Islander',
        '4' => 'No, Neither Aboriginal nor Torres Strait Islander',
        '@' => 'Not stated (Question asked of the client but no answer provided)'
    ];

    /**
     * @var array
     */
    protected static $proficiencyInSpokenEnglish = [
        '1' => 'Very well',
        '2' => 'Well',
        '3' => 'Not well',
        '4' => 'Not at all',
        '@' => 'Not stated',
    ];

    /**
     * @var array
     */
    protected static $disabilityTypes = [
        '11' => 'Hearing/Deaf',
        '12' => 'Physical',
        '13' => 'Intellectual',
        '14' => 'Learning',
        '15' => 'Mental Illness',
        '16' => 'Acquired Brain Impairment',
        '17' => 'Vision',
        '18' => 'Medical Condition',
        '19' => 'Other',
        '99' => 'Unspecified'
    ];
}
